### PR TITLE
TPROD-319 - LeadBundle - Import

### PR DESCRIPTION
**Pull request by JoshuaEstes**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Check that everything relating to importing works - contacts and companies, with all the field matching, segment assignment etc. Check browser and cron based imports.

- Import contacts list is rendered correctly (with and without import history)
- Import contacts list can be sorted
- Import details page is rendered correctly
- The import form is rendered correctly
- Importing the file in the browser will display the same way as it does on the 5.x branch
- Trying to import without adding a csv file will display an error

### CODE REVIEW

File: app/bundles/CoreBundle/Templating/Twig/Extension/DateExtension.php
Comment (escopecz) [inline app/bundles/CoreBundle/Templating/Twig/Extension/DateExtension.php]: I'd remove the second param as it's unused.
```suggestion
    public function formatRange(\DateInterval $range): string
```
Although by looking at the usage of the formatRange method in the EmailBundle there is a "timestamp" sent to it which is obviously a bug that would be discovered by a proper type. But fixing that is not the point of this PR. Please add it to the method comment like `* @param \DateInterval $range` instead.

File: app/bundles/CoreBundle/Templating/Twig/Extension/TranslatorExtension.php
Comment (escopecz) [inline app/bundles/CoreBundle/Templating/Twig/Extension/TranslatorExtension.php]:
```suggestion
    /**
     * Checks for $preferred string existence and returns translation if it does.  Otherwise, returns translation for
     * $alternative.
     *
     * @param array<mixed> $parameters
     */
    public function translatorConditional(string $preferred, string $alternative, array $parameters = [], ?string $domain = null, ?string $locale = null): string
```

File: app/bundles/CoreBundle/Views/Helper/details.html.twig
Comment (escopecz) [inline app/bundles/CoreBundle/Views/Helper/details.html.twig]: Same question here as in the other PR. This shouldn't be necessary according to the twitter docs.

File: app/bundles/LeadBundle/Views/Import/details.html.twig
Comment (escopecz) [inline app/bundles/LeadBundle/Views/Import/details.html.twig]:
```suggestion
```

File: app/bundles/LeadBundle/Views/Import/details.html.twig
Comment (escopecz) [inline app/bundles/LeadBundle/Views/Import/details.html.twig]: This should be resolved

**Review by escopecz — CHANGES_REQUESTED** [review]
Please check the failing CI. I also noticed this output in the failing test HTML:
```html
<li class="last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last nav-group">
```

File: app/bundles/CoreBundle/Views/Helper/details.html.twig
Comment (JoshuaEstes) [reply app/bundles/CoreBundle/Views/Helper/details.html.twig]: Unfortunately. Without it, it will throw an exception. Root cause is https://github.com/mautic/mautic/blob/5.x/app/bundles/CoreBundle/Entity/CommonEntity.php#L36-L45 I figured that using `method_exists` is ok since this is what it does in that method.

This is the same issue with "getPublishUp" as well.

I'm open to another solution as I'm not a huge fan of this either.  I believe this is part of the "strict" being set to true. If this was setting was false, it should work as expected. Reference: https://github.com/twigphp/Twig/blob/3.x/src/Extension/CoreExtension.php#L1604-L1613

File: app/bundles/CoreBundle/Templating/Twig/Extension/TranslatorExtension.php
Comment (JoshuaEstes) [reply app/bundles/CoreBundle/Templating/Twig/Extension/TranslatorExtension.php]: https://github.com/mautic/mautic/pull/11819/files#diff-c9c1c92e357608bf1aa4ca90692c02fcaf55691607e75529006b3bd74f9244c4 This is shared with a few of the PRs

**Comment by JoshuaEstes** [comment]
> Please check the failing CI. I also noticed this output in the failing test HTML:
> 
> ```
> <li class="last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last last nav-group">
> ```

I see that too, I'm not exactly sure what's going on since when it's in dev/prod it doesn't seem to do it.

---

I looked into the failing test and not sure what's going on. When I'm in dev/prod and try to do what the test does in my browser, it works as expected (see screenshot). I've tried a few things and at this point, I'm not sure where to go from here so any advice or guidance is appreciated.

Screenshot 2022-12-16 at 15-43-57 Import Contacts Mautic

File: app/bundles/CoreBundle/Views/Helper/details.html.twig
Comment (escopecz) [reply app/bundles/CoreBundle/Views/Helper/details.html.twig]: Ah, the *magic*. Let's keep it in then.

**Review by mabumusa1 — COMMENTED** [review]
I tested the following screens worked for 

1. Import CSV from the browser worked as expected and values were mapped - Contacts 
2. Import CSV from the browser and assign a segment, worked as expected - Contacts 
3. Import Contacts from CLI worked as expected - Contacts
4. Import Companies from both CLI and web worked as expected.

@JoshuaEstes Please review the issues in the tests, I will confirm everything and approve this PR, thank you for your contribution

**Comment by JoshuaEstes** [comment]
@mabumusa1 That's what I've been trying to figure out. It's very odd because doing the test manually, you can see the error, but the automated test will still fail and, as far as I can tell, there aren't any errors. It's like the form only submits and is validated in the dev/prod environments.

I'm open to ideas or places to explore trying to figure this out.

**Comment by JoshuaEstes** [comment]
@mabumusa1 https://github.com/mautic/mautic/pull/11814/commits/de7409957fa0a829f628497465edba87019785cd resolves the automated test

File: app/bundles/LeadBundle/Views/Import/_list_row.html.twig
Comment (escopecz) [inline app/bundles/LeadBundle/Views/Import/_list_row.html.twig]: I think that the constants must contain the full class where the constant is defined. Otherwise the publish switch is shown all the time.
On 5.x branch:
Screen Shot 2023-01-04 at 17 34 37


On this branch:
Screen Shot 2023-01-04 at 17 39 24

**Review by escopecz — DISMISSED** [review]
I noticed that the double quote sign is being html-encoded in the import form:
Screen Shot 2023-01-04 at 17 26 27

**Review by mabumusa1 — DISMISSED** [review]
I tested this PR again, it looks good to me.  I ran the same tests I ran last time and everything worked as expected.

**Comment by RCheesley** [comment]
@JoshuaEstes please can you attend to the outstanding feedback on this PR? Thanks!

**Review by daniellord32 — CHANGES_REQUESTED** [review]
- [x] contacts import
- [x] contact with segment
- [x] cli import

Just a tiny thing but the info panel needs a text adjustment, %size%MB
image

**Comment by JoshuaEstes** [comment]
@daniellord32 Good catch, the branch has been updated. Let me know if there is anything else you find.

**Review by mabumusa1 — DISMISSED** [review]
I tested the same tests I did last time, this PR is good to go

**Comment by mabumusa1** [comment]
@JoshuaEstes this is almost over, can you please resolve the conflicts?


@RCheesley I think @daniellord32 did a test before, so we just need to resolve the conflicts and then this PR is ready to go

**Comment by RCheesley** [comment]
Yeah unfortunately as we start merging it's more likely we will have to deal with merge conflicts. Hopefully we can start to get these moving more quickly though!

**Comment by RCheesley** [comment]
I've marked this as RTC once the conflicts are resolved - but please make sure we do run a final check over after the conflicts are fixed before merging, just to be sure we don't break anything!

**Comment by JoshuaEstes** [comment]
@RCheesley Conflicts resolved

**Review by mabumusa1 — DISMISSED** [review]
This PR is good to go, it was tested by several contributors see the discussions

**Comment by jos0405** [comment]
Just a small issue:
After import, you can display the contacts imported in the import history by clicking on the number of imported items:
image

This supposed to bring up contacts with this filer:
`import_id:1 import_action:inserted`

But brings up:
`import_id:1 import_id:1 import_action:inserted`

**Comment by jos0405** [comment]
Interrupting import "finish in background" doesn't seem to work.
Reproduce: click on Finish in the background, and the button, and error comes up.
image
image

**Comment by RCheesley** [comment]
@JoshuaEstes @mabumusa1 please can you take a look at the feedback - thanks for the thorough testing @jos0405 🙇

**Comment by mabumusa1** [comment]
> Interrupting import "finish in background" doesn't seem to work. Reproduce: click on Finish in the background, and the button, and error comes up. image image

@jos0405 sometime Gitpod terminates the containers after a period of time if you close the browser menu, keep gitpod opened and then test

**Comment by jos0405** [comment]
FYI: After closing the window (Not pushing cancel, just closing window)
It won't be finished in the background. (I'm not sure this is expected behaviour)
image

**Comment by jos0405** [comment]
> > Interrupting import "finish in background" doesn't seem to work. Reproduce: click on Finish in the background, and the button, and error comes up. image image
> 
> @jos0405 sometime Gitpod terminates the containers after a period of time if you close the browser menu, keep gitpod opened and then test

Yeah, window was open.

**Comment by jos0405** [comment]
> Just a small issue: After import, you can display the contacts imported in the import history by clicking on the number of imported items: image
> 
> This supposed to bring up contacts with this filer: `import_id:1 import_action:inserted`
> 
> But brings up: `import_id:1 import_id:1 import_action:inserted`

FYI: After importing a second file I couldn't reproduce this...

**Comment by jos0405** [comment]
Maybe it's connected
When you click here:
image

The link looks like this:
`/index_dev.php/s/lead/import/view/2`
Gives you an:
image
If the import is still running

> > > Interrupting import "finish in background" doesn't seem to work. Reproduce: click on Finish in the background, and the button, and error comes up. image image
> > 
> > 
> > @jos0405 sometime Gitpod terminates the containers after a period of time if you close the browser menu, keep gitpod opened and then test
> 
> Yeah, window was open.

Maybe it's connected
If the import is not closed yet, and you click here:
image

The link looks like this:
`/index_dev.php/s/lead/import/view/2`
Gives you an:
image
If the import is still running

Once the import is stopped, it looks okay.

**Comment by jos0405** [comment]
In Mautic 4 the import summary page looked like this:
image

In Mautic 5 there are quotation marks around. I actually like it better how in M5 looks. Just wanted to mention the difference:
image

**Comment by mabumusa1** [comment]
@jos0405 I noticed `500 Internal Server Error` when I was testing other PRs, I think `ddev` or `gitpod` goes done during the test for some reason (could be closing the window). 

I just re-tested the whole import process in the background and the UI and it worked in both with any of the issues you mentioned . 

I do not think the error you are seeing now is related to this PR

**Comment by JoshuaEstes** [comment]
While on the 5.x branch and this branch, I get the same issues

---

Given
- A file with 1000 contacts
- Import form is set limit to 100
- All fields are mapped correctly
When
- File is being processed in the browser
Then
- It will ignore the limit and continue to import the entire file

Given
- A file of contact is successfully imported
When
- I view the details of a contact who was imported
- Click on the link to see the import details
Then
- It will return a 500 error

NOTES:
* The 500 error is caused by the route being a 404. The URL should be `/s/contacts/import/view/{id}` however it is `/s/leads/import/view/{id}`. It looks like the link is part of the event, however I haven't spent a lot of time tracking this bug down.
* Reference file https://github.com/mautic/mautic/blob/5.x/app/bundles/LeadBundle/Views/Timeline/list.html.php#L72

Given
- I am on the page that is processing an import
When
- I click "Finish in the background"
Then
- The import will not stop and will give a 500 error in the ajax request

NOTES:
* Resolved

**Comment by jos0405** [comment]
> Import form is set limit to 100
That is just the batch limit. It supposed to ALWAYS import the whole file - unless it is interrupted to continue in the background.

**Comment by RCheesley** [comment]
@jos0405 are all your issues now resolved with this PR or is there something outstanding?

**Review by RCheesley — CHANGES_REQUESTED** [review]
A couple of issues:

`/index_dev.php/s/contacts/import/view/1 `should be the URL for the path in the contacts activity stream, when it's incorrect it's not using the Mautic error screen with our dear friend the Mautibot - not sure why.  It's going to `/index_dev.php/s/lead/import/view/1` as mentioned in earlier reviews, so this doesn't seem to be fixed yet.

When I clicked this button it went to just the list of all contacts, not a filtered list of contacts imported by this segment.  The path it should use is `index_dev.php/s/contacts?search=import_id:<number of import>`.

screenshot-8080-joshuaestes-mautic-hxbl901jauw ws-eu85 gitpod io-2023 02 08-13_22_11

Otherwise the import worked OK, all the links in the import list work, and the import overview works.

**Comment by JoshuaEstes** [comment]
@RCheesley 

> `/index_dev.php/s/contacts/import/view/1 `should be the URL for the path in the contacts activity stream, when it's incorrect it's not using the Mautic error screen with our dear friend the Mautibot - not sure why. It's going to `/index_dev.php/s/lead/import/view/1` as mentioned in earlier reviews, so this doesn't seem to be fixed yet.

https://github.com/mautic/mautic/pull/11814/commits/07bd692ca81fcbebf537ca3c150868698c46088a This bug exists on 5.x branch as well and this will resolve that issue.

> When I clicked this button it went to just the list of all contacts, not a filtered list of contacts imported by this segment. The path it should use is `index_dev.php/s/contacts?search=import_id:<number of import>`.

https://github.com/mautic/mautic/pull/11814/commits/789203d7e81b05024d33f32330c2ce17c8d5a946 Resolved

**Comment by mabumusa1** [comment]
@RCheesley this PR is pending your approval on the changes, when you have the chance please test so we can merge it

**Review by jos0405 — APPROVED** [review]
I retested the import behaviour, looks good!
All my issues were addressed.

**Review by RCheesley — APPROVED** [review]

**Pull request merged**
